sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryHashPartitionVerifySuite.scala

```diff
@@ -0,0 +1,219 @@
+        .toSeq
+    } finally source.close()
+  }
+
+  private def saveRowsAndPartIds(rows: Seq[UnsafeRow], partIds: Seq[Int], os: DataOutputStream) = {
+    // Save the total number of rows
+    os.writeInt(rows.length)
+    // Save all rows
+    rows.foreach { row =>
+      // Save the row's total number of bytes
+      val rowBytes = row.getBytes()
+      // Save the row's actual bytes
+      os.writeInt(rowBytes.size)
+      os.write(rowBytes)
+    }
+    // Save all partIds, which should be in the same order as rows
+    partIds.foreach { id =>
+      os.writeInt(id)
+    }
+  }
+
+  private def readRowsAndPartIds(is: DataInputStream): (Seq[UnsafeRow], Seq[Int]) = {
+    val numRows = is.readInt()
+    val rows = (1 to numRows).map { _ =>
+      val rowSize = is.readInt()
+      val rowBuffer = new Array[Byte](rowSize)
+      ByteStreams.readFully(is, rowBuffer, 0, rowSize)
+      val row = new UnsafeRow(1)
+      row.pointTo(rowBuffer, rowSize)
+      row
+    }
+    val partIds = (1 to numRows).map(_ => is.readInt()).toArray.toSeq
+    (rows, partIds)
+  }
+
+  private def getRandomRows(schema: StructType, numRows: Int, rand: Random): Seq[UnsafeRow] = {
+    val generator = RandomDataGenerator
+      .forType(
+        schema,
+        rand = new Random(rand.nextInt())
+      )
+      .get
+
+    // Create the converters needed to convert from external row to internal
+    // row and to UnsafeRows. Projection itself costs a lot on initialization
+    // (codegen and compile), so initialize it once.
+    val internalConverter = CatalystTypeConverters.createToCatalystConverter(schema)
+    val unsafeConverter = UnsafeProjection.create(Array(schema).asInstanceOf[Array[DataType]])
+
+    (1 to numRows).map { _ =>
+      val row = generator().asInstanceOf[Row]
+
+      val internalRow = new GenericInternalRow(1)
+      internalRow.update(0, internalConverter(row).asInstanceOf[GenericInternalRow])
+      val unsafeRow = unsafeConverter.apply(internalRow)
+
+      // UnsafeProjection returns the same UnsafeRow instance intentionally, so
+      // unless doing deep copy, the hash partitions below will evaluate on the
+      // same row thus return same value.
+      unsafeRow.copy()
+    }
+  }
+
+  private def getRandomSchemas(rand: Random): Seq[StructType] = {
+    (1 to numSchemaTypes).map { _ =>
+      RandomDataGenerator.randomNestedSchema(rand, maxNumFields, variableTypes)
+    }
+  }
+
+  private def getPartitionId(rows: Seq[UnsafeRow], hash: HashPartitioning): Seq[Int] = {
+    val partIdExpr = hash.partitionIdExpression
+    rows.map { row =>
+      partIdExpr.eval(row).asInstanceOf[Int]
+    }
+  }
+
+  test("SPARK-47788: Ensure the same hash function is used across batches.") {
+    val rowAndPartIdFile = new File(goldenFileURI.getPath, rowAndPartIdFilename)
+
+    if (regenerateGoldenFiles) {
+      // To limit the golden file size under 10Mb, please set the final val MAX_STR_LEN: Int = 100
```

Comment:
Do I understand correctly that RandomDataGenerator.forType receives the parameters for max string length and max array size? We don't need to let others to fix the code of RandomDataGenerator. Let's just do that by ourselves via providing the parameters.

Reply:
In the oss-spark, it doesn't take parameters for max string length and max array size, which differs databricks' internal version.